### PR TITLE
GitHub release

### PR DESCRIPTION
**Pull request by badboy**
This is a first shot at #15 and #16 

* [x] Push to GitHub
    * [x] Make sure it works on Travis with proper git key
* [x] Create release on GitHub
    * [x] Better user key handling
* [x] Publish on crates.io

### CODE REVIEW

**Comment by badboy** [comment]
I'm working on getting some patches into hubcaps so we can easily handle their errors.

**Comment by badboy** [comment]
We need to extract the github repository (`user/repo`) and execute `git push https://TOKEN@github.com/user/repo.git`

**Comment by badboy** [comment]
Ah, and of course it breaks the tests because it assumes it can push and create releases (same will happen for crates.io publishing).
Should we have a flag to explicitely disable the last steps?

**Comment by badboy** [comment]
Force-pushed the rebased stuff.

**Comment by schultyy** [comment]
I guess it makes sense to opt out or better opt in for publishing new releases on GitHub. Doing it all automatically might not fit in every use case.

**Comment by badboy** [comment]
Yup, my idea was opt out. The default is debatable. I will implement it and we can see what we think is better

**Comment by schultyy** [comment]
:+1:

**Comment by badboy** [comment]
nightly tests broken because compiler bug

File: src/git.rs [hunk 1]
```diff
@@ -117,25 +116,51 @@ pub fn generate_commit_message(new_version: &str) -> String {
...
+    let repo_name = &config.repository_name;
+    let branch    = &config.branch;
+
+    // We need to push both the branch we just committed as well as the tag we created.
+    let branch_ref = format!("refs/heads/{}", branch);
+    let tag_ref    = format!("refs/tags/{}", tag_name);
+    let refs = [&branch_ref[..], &tag_ref[..]];
 
-    create_tag(&repo, &signature, &tag_name, &tag_message)
+    // TODO: Get user and repo
```
Comment (schultyy) [inline src/git.rs]: Is here still a Todo or is that done?

File: src/github.rs [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+use config::Config;
+
+pub fn release(config: &Config, tag_name: &str, tag_message: &str) -> Result<(), Error> {
+    let user      = &config.user[..];
+    let repo_name = &config.repository_name[..];
+    let branch    = &config.branch[..];
+    let token     = config.gh_token.as_ref().unwrap();
+
+    let client = Client::new();
+    let github = Github::new(USERAGENT, &client, Some(&token[..]));
```
Comment (schultyy) [inline src/github.rs]: What's `USERAGENT` here? Is it the application name?

File: src/main.rs [hunk 3]
```diff
@@ -75,11 +95,36 @@ fn ci_env_set() -> bool {
     env::var("CI").is_ok()
 }
 
+fn user_repo_from_url(url: Url) -> Result<(String, String), String> {
```
Comment (schultyy) [inline src/main.rs]: What does `(String, String)` mean?

File: src/git.rs (hunk 1 above)
Comment (badboy) [inline src/git.rs]: Done. I'll remove that

File: src/github.rs (hunk 2 above)
Comment (badboy) [inline src/github.rs]: I defined it in main: https://github.com/semantic-rs/semantic-rs/blob/github-release/src/main.rs#L37
```rust
const USERAGENT: &'static str = concat!("semantic-rs/", env!("CARGO_PKG_VERSION"));
```

It's passed as the HTTP useragent when sending requests to GitHub

File: src/main.rs [hunk 4]
```diff
@@ -75,11 +95,36 @@ fn ci_env_set() -> bool {
...
+    let user = path[0].clone();
+    let repo = match path[1].rfind(".git") {
+        None => path[1].clone(),
+        Some(suffix_pos) => {
+            let valid_pos = path[1].len() - 4;
+            if valid_pos == suffix_pos {
+                let path = &path[1][0..suffix_pos];
+                path.into()
+            } else {
+                return Err(".git suffix not at the end of URL".into())
```
Comment (schultyy) [inline src/main.rs]: What about `URL does not point to a git repository (Example: http://example.com/foo.git)`?

File: src/main.rs (hunk 3 above)
Comment (badboy) [inline src/main.rs]: A tuple containing two strings. This allows for multiple return values and we extract both user and repo name at the same time

File: src/main.rs (hunk 4 above)
Comment (badboy) [inline src/main.rs]: As the error message? I like it much more

File: src/main.rs [hunk 5]
```diff
@@ -127,11 +182,44 @@ If none is set the normal git config is tried in the following order:
...
+        let cargo_token = env::var("CARGO_TOKEN")
+            .unwrap_or_else(|err| print_exit!("CARGO_TOKEN not set: {:?}", err));
 
-    let version = toml_file::read_from_file(repository_path)
+        cb.gh_token(gh_token);
+        cb.cargo_token(cargo_token);
+    }
+
+    cb.repository(repo);
+    cb.branch("master".into());
```
Comment (schultyy) [inline src/main.rs]: Do we use `master` only right now?

Comment (badboy) [inline src/main.rs]: Actually we have a flag for that. This is a left-over from testing. Will use `args.flag_branch` instead (which defaults to master)

**Pull request merged**